x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/datatiers/DataTiersUsageTransportAction.java

```diff
@@ -0,0 +1,205 @@
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureAction;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureResponse;
+import org.elasticsearch.xpack.core.action.XPackUsageFeatureTransportAction;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+public class DataTiersUsageTransportAction extends XPackUsageFeatureTransportAction {
+
+    private final Client client;
+
+    @Inject
+    public DataTiersUsageTransportAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Client client
+    ) {
+        super(
+            XPackUsageFeatureAction.DATA_TIERS.name(),
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            indexNameExpressionResolver
+        );
+        this.client = client;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        XPackUsageRequest request,
+        ClusterState state,
+        ActionListener<XPackUsageFeatureResponse> listener
+    ) {
+        new ParentTaskAssigningClient(client, clusterService.localNode(), task).admin()
+            .cluster()
+            .execute(
+                NodesDataTiersUsageTransportAction.TYPE,
+                new NodesDataTiersUsageTransportAction.NodesRequest(),
+                listener.delegateFailureAndWrap((delegate, response) -> {
+                    // Generate tier specific stats for the nodes and indices
+                    delegate.onResponse(
+                        new XPackUsageFeatureResponse(
+                            new DataTiersFeatureSetUsage(
+                                aggregateStats(response.getNodes(), getIndicesGroupedByTier(state, response.getNodes()))
+                            )
+                        )
+                    );
+                })
+            );
+    }
+
+    // Visible for testing
+    static Map<String, Set<String>> getIndicesGroupedByTier(ClusterState state, List<NodeDataTiersUsage> nodes) {
+        Set<String> indices = nodes.stream()
+            .map(nodeResponse -> state.getRoutingNodes().node(nodeResponse.getNode().getId()))
+            .filter(Objects::nonNull)
+            .flatMap(node -> StreamSupport.stream(node.spliterator(), false))
+            .map(ShardRouting::getIndexName)
+            .collect(Collectors.toSet());
+        Map<String, Set<String>> indicesByTierPreference = new HashMap<>();
+        for (String indexName : indices) {
+            List<String> tierPreference = state.metadata().index(indexName).getTierPreference();
```

Comment:
I think we should add a NPE check here as by the time the nodes reported their usages, some indices might've been deleted from the cluster state.
`state.metadata().index(indexName)` can be null.

Reply:
Ah! Good catch!